docs/tutorial.rst

```diff
@@ -306,17 +318,303 @@ trees. In this example we incorrectly inferred that 0 coalesces with the
+``add_diploid_sites`` illustrates how to iterate over the variants in a ``CYVCF2.VCF``
+object and add them to a ``tsinfer`` sample data file, while printing out the progression
+through the file in number of base pairs. This particular script assumes that the VCF
+describes a set of phased, diploid individuals, and has the ancestral state present in
+the INFO field of the VCF (otherwise it takes the reference allele as the ancestral
+state). For example data, we use a publicly available VCF file of the genetic variants
+from chromosome 24 of ten Norwegian and French house sparrows, *Passer domesticus*
+(thanks to Mark Ravinet for the data file):
 
-    1. Add documentation links for msprime above so we can explain tree
-       sequences there.
+.. code-block:: python
 
-++++++++++++
-Data example
-++++++++++++
+    import cyvcf2
+    import tsinfer
+    import tqdm
+
+    def add_diploid_sites(vcf, samples):
+        """
+        Loop over sites in the vcf and add them to the samples object, reordering the
+        alleles to put the ancestral allele first, if it is available.
+        """
+        progressbar = tqdm.tqdm(total=samples.sequence_length, desc="Read VCF", unit='bp')
+        pos = 0
+        for variant in vcf:
+            if pos and pos == variant.POS:
+                raise ValueError("Duplicate positions for variant at position", pos)
+            else:
+                progressbar.update(variant.POS - pos)
+                pos = variant.POS
+            if any([not phased for _, _, phased in variant.genotypes]):
+                raise ValueError("Unphased genotypes for variant at position", pos)
+            alleles = [variant.REF] + variant.ALT
+            # Sort with the ancestral allele first if we have one (otherwise use REF)
+            ancestral = variant.INFO.get('AA', variant.REF)
+            all_alleles = [variant.REF] + variant.ALT
+            # Ancestral state must be first in the allele list.
+            alleles = [ancestral] + list(set(all_alleles) - {ancestral})
+            allele_index = {old_index: alleles.index(allele)
+                for old_index, allele in enumerate(all_alleles)}
+            # Map original allele indexes to their indexes in the new alleles list.
+            genotypes = [allele_index[old_index]
+                for row in variant.genotypes for old_index in row[0:2]]
+            samples.add_site(pos, genotypes=genotypes, alleles=alleles)
+        progressbar.close()
+
+    def chromosome_length(vcf):
+        try:
+            if len(vcf.seqlens) > 1:
+                raise ValueError("More than one contig length listed in the VCF file")
+            return vcf.seqlens[0]
+        except (AttributeError, IndexError):
+            print("No chromosome length in VCF file, percent progress hidden", flush=True)
+            return 0
+
+    # URL for the VCF
+    url = ("https://github.com/tskit-dev/tsinfer/raw/master/docs/_static/P_dom_chr24_phased.vcf.gz")
```

Comment:
What are the brackets for?

Reply:
Ah, left in when I had to multi-line the URL for length reasons. At the moment this line is over the 89 char limit (94 chars, actually), but it looks OK in the docs. Do we want to enforce the 89 char limit in Python code in the docs too?

Reply:
94 is fine for the docs I think.